book/src/e2e-testing.md

```diff
@@ -0,0 +1,81 @@
+
+To be able import the in-tree storage tests, CSI plugin authors would need to implement a [testdriver](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L31) for their CSI plugin. The testdriver provides required functionality that would help setup testcases for a particular plugin. 
+
+For any testdriver the basic required functions are (Since it implements the [TestDriver Interface](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L31)):
+ - `GetDriverInfo() *testsuites.DriverInfo`
+ - `SkipUnsupportedTest(pattern testpatterns.TestPattern)`
+ - `PrepareTest(f *framework.Framework) (*testsuites.PerTestConfig, func())` 
+ 
+The `PrepareTest` method is important as it is where you will write code to deploy your driver and setup other pre-requistes that your driver requires. The nfs and hostpath testdrivers deploy the plugin by simply [using the manifests files](https://github.com/mathu97/csi-driver-nfs/blob/master/test/nfs_driver.go#L91) that you would regularly use to deploy a plugin. 
+
+`GetDriverInfo` will return a `DriverInfo` object that has all of the plugin's capabilities and required information. This object helps tests find the driver when it is deployed, and also decides which tests should run (depending on if the plugin supports certain capabilities).
+
+Here are examples of the NFS and Hostpath DriverInfo objects:
+
+```
+testsuites.DriverInfo{
+			Name:        "csi-nfsplugin",
+			MaxFileSize: testpatterns.FileSizeLarge,
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+			Capabilities: map[testsuites.Capability]bool{
+				testsuites.CapPersistence: true,
+				testsuites.CapExec:        true,
+			},
+}
+```
+
+```
+testsuites.DriverInfo{
+			Name:        "csi-hostpath",
+			FeatureTag:  "",
+			MaxFileSize: testpatterns.FileSizeMedium,
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+			Capabilities: map[testsuites.Capability]bool{
+				testsuites.CapPersistence: true,
+			},
+}
+```
+
+You would define something similar for your CSI plugin.
+
+`SkipUnsupportedTest` simply skips any tests that you define there.
+
+Depending on your plugin's specs, you would implement other driver interaces defined [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L61). For example the [NFS testdriver](https://github.com/mathu97/csi-driver-nfs/blob/master/test/nfs_driver.go) also implements PreprovisionedVolumeTestDriver, and PreprovisionedPVTestDriver interfaces, to enable pre-provisoned tests. 
+
+After implementing the tesdriver for your CSI plugin, you would create a `csi-volumes.go` file, where the implemented testdriver is used to run in-tree storage test suites [similar to how the NFS testdriver does](https://github.com/mathu97/csi-driver-nfs/blob/master/test/csi-volumes.go#L38). 
```

Comment:
typo: testdriver

Reply:
Also mention here that this is where you define which test suites you want to run.  And you can find the list of all test suites at __